src/utils/wallet/solana.ts

```diff
@@ -102,125 +108,194 @@ export function fetchFogoOptions() {
+ * @param connection The Solana connection object
+ * @param transaction The transaction to resend
+ */
+
+async function resendTransactionUntilConfirmed(
+  connection: Connection,
+  transaction: {
+    serializedTransaction: Uint8Array | Buffer | number[];
+    sendOptions?: SendOptions;
+  },
+  blockhash: string,
+  lastValidBlockHeight: number,
+  commitment: Commitment,
+): Promise<string> {
+  let isTransactionConfirmed: RpcResponseAndContext<SignatureResult> | null =
+    null;
+
+  const signature = await connection.sendRawTransaction(
+    transaction.serializedTransaction,
+    transaction.sendOptions,
+  );
+
+  const confirmPromise = connection.confirmTransaction(
+    { signature, blockhash, lastValidBlockHeight },
     commitment,
   );
 
   try {
-    // This loop will break once the transaction has been confirmed or the block height is exceeded.
-    // An exception will be thrown if the block height is exceeded by the confirmTransactionPromise.
-    // The transaction will be resent if it hasn't been confirmed after the interval.
-    const txRetryInterval = 5000;
-    while (!confirmedTx) {
-      confirmedTx = await Promise.race([
-        confirmTransactionPromise,
+    while (!isTransactionConfirmed) {
+      isTransactionConfirmed = await Promise.race([
+        confirmPromise,
         new Promise<null>((resolve) =>
-          setTimeout(() => {
-            resolve(null);
-          }, txRetryInterval),
+          setTimeout(() => resolve(null), CONFIRMATION_PROMISE_TIMER),
         ),
       ]);
-      if (confirmedTx) {
+
+      if (isTransactionConfirmed) {
         break;
       }
-      console.log(
-        `Tx not confirmed after ${
-          txRetryInterval * txSendAttempts++
-        }ms, resending`,
+
+      // Resend transaction as it's not yet confirmed
+      await connection.sendRawTransaction(
+        transaction.serializedTransaction,
+        transaction.sendOptions,
       );
-      try {
-        await connection.sendRawTransaction(serializedTx, sendOptions);
-      } catch (e) {
-        console.error('Failed to resend transaction:', e);
-      }
     }
-  } catch (e: unknown) {
+  } catch (e) {
     if (
       e instanceof Error &&
       e.name === 'TransactionExpiredBlockheightExceededError'
     ) {
-      // Sometimes the transaction actually landed,
-      // so spend some additional time to check for it
-      const maxRetries = 5;
-      const retryDelay = 2000;
-      for (let i = 0; i < maxRetries; ++i) {
-        try {
-          const tx = await connection.getTransaction(signature, {
-            commitment: 'confirmed',
-            maxSupportedTransactionVersion: 0,
-          });
-          if (tx) return signature; // Transaction actually landed
-        } catch {
-          // Ignore errors, we will retry
-        }
-        await sleep(retryDelay);
-      }
+      await recoverBlockheightExceededTransaction(connection, signature);
+      return signature;
```

Comment:
We should only return the `signature` if the transaction was actually confirmed. To preserve the functionality from before, we could do:
```
const recoveredSignature = await recoverBlockheightExceededTransaction(connection, signature);
if (recoveredSignature) return recoveredSignature;
throw e;
```
We would then change `recoverBlockheightExceededTransaction` to return Promise<string | null>